tokio-executor/src/thread_pool/park.rs

```diff
@@ -0,0 +1,181 @@
+                return;
+            }
+        }
+
+        // Otherwise we need to coordinate going to sleep.
+        let mut _m = self.lock.lock().unwrap();
+
+        match self.state.compare_exchange(EMPTY, PARKED, SeqCst, SeqCst) {
+            Ok(_) => {}
+            // Consume this notification to avoid spurious wakeups in the next park.
+            Err(NOTIFIED) => {
+                // We must read `state` here, even though we know it will be `NOTIFIED`. This is
+                // because `unpark` may have been called again since we read `NOTIFIED` in the
+                // `compare_exchange` above. We must perform an acquire operation that synchronizes
+                // with that `unpark` to observe any writes it made before the call to `unpark`. To
+                // do that we must read from the write it made to `state`.
+                let old = self.state.swap(EMPTY, SeqCst);
+                assert_eq!(old, NOTIFIED, "park state changed unexpectedly");
+                return;
+            }
+            Err(n) => panic!("inconsistent park_timeout state: {}", n),
+        }
+
+        match timeout {
+            None => {
+                loop {
+                    // Block the current thread on the conditional variable.
+                    _m = self.cvar.wait(_m).unwrap();
+
+                    match self.state.compare_exchange(NOTIFIED, EMPTY, SeqCst, SeqCst) {
+                        Ok(_) => return, // got a notification
+                        Err(_) => {}     // spurious wakeup, go back to sleep
+                    }
+                }
+            }
+            Some(timeout) => {
+                // Wait with a timeout, and if we spuriously wake up or otherwise wake up from a
+                // notification we just want to unconditionally set `state` back to `EMPTY`, either
+                // consuming a notification or un-flagging ourselves as parked.
+                _m = self.cvar.wait_timeout(_m, timeout).unwrap().0;
+
+                match self.state.swap(EMPTY, SeqCst) {
+                    NOTIFIED => {} // got a notification
+                    PARKED => {}   // no notification
+                    n => panic!("inconsistent park_timeout state: {}", n),
+                }
+            }
+        }
+    }
+
+    fn unpark(&self) {
+        // To ensure the unparked thread will observe any writes we made before this call, we must
+        // perform a release operation that `park` can synchronize with. To do that we must write
+        // `NOTIFIED` even if `state` is already `NOTIFIED`. That is why this must be a swap rather
+        // than a compare-and-swap that returns if it reads `NOTIFIED` on failure.
+        match self.state.swap(NOTIFIED, SeqCst) {
+            EMPTY => return,    // no one was waiting
+            NOTIFIED => return, // already unparked
+            PARKED => {}        // gotta go wake someone up
+            _ => panic!("inconsistent state in unpark"),
```

Comment:
Nit: the panic in `park_timeout` records the invalid state; should this also?
```suggestion
            n => panic!("inconsistent state in unpark: {}", n),
```